File: cousinsbt.go
// Time Complexity : O(n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this :

// Your code here along with comments explaining your approach
/*
add root to queue
while q not empty loop till size
for current node if x value or y value is found and set it to true
if xfound or y found is set to true already check if other y or x is found then return true
if node and its children are not nil check if we find both x, y as siblings then return false
append left, right children to queue
*/
package main

import "fmt"

type TreeNode struct {
	Val int
	Left *TreeNode
	Right *TreeNode
}

func isCousins(root *TreeNode, x int, y int) bool {
	if root == nil {
		return false
	}
	q := []*TreeNode{root}
	for len(q) > 0 {
		size := len(q)
		x_found := false
		y_found := false
		for i:=0;i<size;i++ {
			n := q[0]
			q = q[1:]
			if n != nil {
				if n.Val == x {
					x_found = true
				}
				if n.Val == y {
					y_found = true
				}
			}
			if x_found {
				if n != nil && n.Val == y {
					return true
				}
			}
			if y_found {
				if n != nil && n.Val == x {
					return true
				}
			}
			if n != nil && n.Left != nil && n.Right != nil {
				if n.Left.Val == x && n.Right.Val == y {
					return false
				}
				if n.Left.Val == y && n.Right.Val == x {
					return false
				}
			}
			if n != nil && n.Left != nil {
				q = append(q, n.Left)
			}
			if n != nil && n.Right != nil {
				q = append(q, n.Right)
			}
		}
	}
	return false
}

func MainCousin() {
	root := &TreeNode{Val: 1, Left: &TreeNode{Val: 2, Left: &TreeNode{Val:4}}, Right: &TreeNode{Val: 3}}
	fmt.Println(isCousins(root, 4, 3)) //expected false
}

File: CousinsBt.java
//Time Complexity: O(n) 
//Space Complexity: O(h)
// DFS
public class CousinsBt {
    TreeNode x_parent, y_parent;
    int x_level, y_level;
    public boolean isCousins(TreeNode root, int x, int y) {
        if(root == null)
            return false;
        
        dfs(root, x, y, 0, null);
        return x_parent != y_parent && x_level == y_level;
        
    }
    private void dfs(TreeNode root, int x, int y, int lvl, TreeNode parent){
        if(root == null)
            return;
        if(root.val == x){
            x_parent = parent;
            x_level = lvl;
        }
        if(root.val == y){
            y_parent = parent;
            y_level = lvl;
        }
        dfs(root.left, x, y, lvl + 1, root);
        dfs(root.right, x, y, lvl + 1, root);
        
    }
}
//Time Complexity: O(n) 
//Space Complexity: O(h)
// BFS
public class CousinsBt {
    public boolean isCousins(TreeNode root, int x, int y) {
        if(root == null)
            return false;
        
        Queue<TreeNode> q = new LinkedList<>();
        q.offer(root);
        
        while(!q.isEmpty())
        {
            int size = q.size();
            boolean x_found = false, y_found = false;
            for(int i = 0 ; i < size ; i ++)
            {
                TreeNode curr = q.poll();
                if(curr.val == x)
                {
                    x_found = true;
                }
                else if(curr.val == y)
                {
                    y_found = true;
                }
                
                if(x_found == true && y_found == true)
                    return true;
                if(x_found || y_found)
                    continue;
                // same parent
                if(curr.left != null && curr.right != null && curr.left.val == x && curr.right.val == y)                                                return false;
                // same parent
                if(curr.left != null && curr.right != null && curr.left.val == y && curr.right.val == x)
                    return false;
                if(curr.left != null)
                    q.offer(curr.left);
                if(curr.right != null)
                    q.offer(curr.right);
            }
            if(x_found == true && y_found == true)
                return true;
            if(x_found || y_found)
                return false;
        }
        return false;
    }
}
